src/toxcore/dht/ping_array.rs

```diff
@@ -0,0 +1,206 @@
+/*
+    Copyright © 2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Manage ping_id.
+Generate ping_id on request packet, check ping_id on response packet.
+*/
+
+use std::io::{ErrorKind, Error};
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+
+use toxcore::crypto_core::*;
+use toxcore::dht::packed_node::*;
+
+/// Ping timeout in seconds
+pub const PING_TIMEOUT: u64 = 5;
+
+/// size of DHT ping arrays.
+pub const DHT_PING_ARRAY_SIZE: u64 = 512;
+
+/// Entry type of PingArray
+#[derive(Clone, Debug)]
+struct PingArrayEntry {
+    data: Vec<PackedNode>,
+    time: Instant,
+}
+
+/// Generate ping_id, check if ping_id is correct or timedout
+#[derive(Clone, Debug)]
+pub struct PingArray {
+    entries: HashMap<u64, PingArrayEntry>,
+    last_added: u64,
+    total_size: u64,
+    timeout: Duration,
+}
+
+impl PingArray {
+    /// Create new PingArray object
+    pub fn new(total_size: u64, timeout: Duration) -> Option<PingArray> {
+        if total_size == 0 {
+            return None;
+        }
+
+        Some(PingArray {
+            entries: HashMap::new(),
+            last_added: 0,
+            total_size,
+            timeout,
+        })
+    }
+
+    // Clear timed out entries
+    fn ping_array_clear_timedout(&mut self) {
+        let timedout_dur = self.timeout;
```

Comment:
Why do you strore it in struct instead of passing as a parameter?

Reply:
self.timeout setup is done when PingArray::new(timeout) as a parameter.
Each PingArray object has same timeout value.
When PingArray for NodesReq/Res it's timeout is 5 seconds and 5 seconds is passed as a parameter.
When PingArray for other Req/Res it's timeout setting is available using parameter.

Reply:
`ping_array_clear_timedout` is a very heavy operation. it is `O(N)`. So calling it in `ping_array_add` makes `ping_array_add` method very heavy too. It would be better to add periodic tasks into example/dht_node.rs as it was done for timedout clients and remove timedouts from PingArray for NodesReq/Res and from PingArray for other Req/Res each second.

Reply:
Also timeout duration as arg will help you to test your code!

Reply:
Good idea.